clash-lib/src/Clash/Netlist/Id.hs

```diff
@@ -90,35 +90,40 @@ import qualified Clash.Netlist.Id.Common as Common
 emptyIdentifierSet
   :: Bool
   -- ^ Allow escaped identifiers?
+  -> Bool
+  -- ^ Should all basic identifiers be lower case?
   -> HDL
   -- ^ HDL to generate names for
   -> IdentifierSet
-emptyIdentifierSet esc hdl = makeSet esc hdl []
+emptyIdentifierSet esc lw hdl = makeSet esc lw hdl []
 
 -- | Union of two identifier sets. Errors if given sets have been made with
 -- different options enabled.
 union :: HasCallStack => IdentifierSet -> IdentifierSet -> IdentifierSet
-union is1@(IdentifierSet esc1 hdl1 _ _) is2@(IdentifierSet esc2 hdl2 _ _)
+union is1@(IdentifierSet esc1 lw1 hdl1 _ _) is2@(IdentifierSet esc2 lw2 hdl2 _ _)
   | esc1 /= esc2 = error $ "Internal error: esc1 /= esc2, " <> show (esc1, esc2)
   | hdl1 /= hdl2 = error $ "Internal error: hdl1 /= hdl2, " <> show (hdl1, hdl2)
-  | otherwise = makeSet esc1 hdl1 (toList is1 ++ toList is2)
+  | lw1 /= lw2 = error $ "Internal error: lw1 /= lw2 , " <> show (lw1, lw2)
+  | otherwise = makeSet esc1 lw1 hdl1 (toList is1 ++ toList is2)
 
 -- | Make a identifier set filled with given identifiers
 makeSet
   :: Bool
   -- ^ Allow escaped identifiers?
+  -> Bool
```

Comment:
Perhaps use a custom enum here:
```haskell
data PreserveCase
  = PreserveCase
  | ToLower
```

And of course propagate this throughout the rest of the design. Just so that we have fewer "random" Booleans in our codebase.